allennlp_rc/models/transformer_qa.py

```diff
@@ -0,0 +1,233 @@
+            A tensor of shape ``(batch_size, passage_length)`` representing unnormalized log
+            probabilities of the span end position (inclusive).
+        span_end_probs : torch.FloatTensor
+            The result of ``softmax(span_end_logits)``.
+        best_span : torch.IntTensor
+            The result of a constrained inference over ``span_start_logits`` and
+            ``span_end_logits`` to find the most probable span.  Shape is ``(batch_size, 2)``
+            and each offset is a token index.
+        best_span_scores : torch.FloatTensor
+            The score for each of the best spans.
+        loss : torch.FloatTensor, optional
+            A scalar loss to be optimised.
+        best_span_str : List[str]
+            If sufficient metadata was provided for the instances in the batch, we also return the
+            string from the original passage that the model thinks is the best answer to the
+            question.
+        """
+        embedded_question = self._text_field_embedder(question)
+        logits = self._linear_layer(embedded_question)
+        span_start_logits, span_end_logits = logits.split(1, dim=-1)
+        span_start_logits = span_start_logits.squeeze(-1)
+        span_end_logits = span_end_logits.squeeze(-1)
+
+        possible_answer_mask = torch.zeros_like(question["tokens"]["mask"])
+        for i, (start, end) in enumerate(context_span):
+            possible_answer_mask[i, start : end + 1] = 1
+
+        span_start_logits = util.replace_masked_values(
+            span_start_logits, possible_answer_mask, -1e32
+        )
+        span_end_logits = util.replace_masked_values(span_end_logits, possible_answer_mask, -1e32)
+        span_start_probs = torch.nn.functional.softmax(span_start_logits, dim=-1)
+        span_end_probs = torch.nn.functional.softmax(span_end_logits, dim=-1)
+        best_spans = get_best_span(span_start_logits, span_end_logits)
+        best_span_scores = torch.gather(
+            span_start_logits, 1, best_spans[:, 0].unsqueeze(1)
+        ) + torch.gather(span_end_logits, 1, best_spans[:, 1].unsqueeze(1))
+        best_span_scores = best_span_scores.squeeze(1)
+
+        output_dict = {
+            "span_start_logits": span_start_logits,
+            "span_start_probs": span_start_probs,
+            "span_end_logits": span_end_logits,
+            "span_end_probs": span_end_probs,
+            "best_span": best_spans,
+            "best_span_scores": best_span_scores,
+        }
+
+        # Compute the loss for training.
+        if answer_span is not None:
+            span_start = answer_span[:, 0]
+            span_end = answer_span[:, 1]
+            span_mask = span_start != -1
+            self._span_accuracy(
+                best_spans, answer_span, span_mask.unsqueeze(-1).expand_as(best_spans)
+            )
+
+            start_loss = cross_entropy(span_start_logits, span_start, ignore_index=-1)
+            if torch.any(start_loss > 1e9):
```

Comment:
Might be useful to pull this out into a single method call somewhere.  It reads like it's a bunch of print statements for debugging purposes.  Or you could just leave it out of the code and only add it when you're debugging.

Reply:
It's fairly handy. When you are working on a different reader to feed this model (to read SQuAD 2.0, or some other data set), you might get the annotated offsets wrong. This is where you find out first, without training for a long time on a doomed model. If it's our own reader, maybe you want to say "do it in a test" instead, but not all of them are going to be our own, and even then it's helpful to have this.

There was a typo in the second block, which I fixed in bc55eef910dbce5cdf65188d77bebfa399c9ece8.